### ISSUE TITLE
False error 7022, circular reference when there is actually none

### ISSUE THREAD

**Issue by garyo** [microsoft/TypeScript#43708]
# Bug Report

### 🕗 Version & Regression Information

This is new code, but I'm using 3.8.3. Still occurs in 4.2.3. I asked in the Discord and they said it looks like a bug.

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/C4TwDgpgBAqgzhATgJQgYwPaICZQLxQDeAdgIYC2EAXFHMIgJbEDmAvgFClwjFpQBmAV17AGGYlAA2DOvCRwAFMAwBrCMRp1GLAD7DsEfkwjYAlEXZQrURBGCDEEwoISI4NANrOG2GgEYAGigyShoAcgBxUkQQMNYAXSCwUmYIABVVdXCAIzj2DnYIAA8wLGAoLh4+IRExCWk6AEFBYAALOTcFc0JLKTtg4uAABRT0zIkCfUNjbF7MYjooF3kaQg9BH016JmZ4mg7UTBxWfCIOK2wMC2soecXltwAZGXKCUgB3UgZyhuAOxWIgxGqQyamIpl6VgecGedAAdNC4fwsABRUhoVoKBQPcx4AB81xuS1ccHWrgRPnipwekKgrAhN0BRWGo1B6mpJNhwDhyRB416J3erQYkmgCiZLL5YIhrCAA)

### 💻 Code

```ts
type UserRecord = {name: string}
async function listUsers(token: string|undefined) {
    return {users: [{uid: 1, name: 'Gary'}], pageToken: 'b'}
}

export async function listAuthUsers() {
  let nextPageToken = undefined
  const users: {[uid: string]: UserRecord} = {}
  do {
    const usersList = await listUsers(nextPageToken) // <<<< ERROR 7022 here
    usersList.users.forEach((user) => {
      users[user.uid] = user
    })
    nextPageToken = usersList.pageToken // <<< ERROR DISAPPEARS IF THIS LINE IS COMMENTED OUT
  } while (nextPageToken)
}
```

### 🙁 Actual behavior

Code above gives error 7022. If you comment out the line `nextPageToken = usersList.pageToken ` then it's OK.

User webstrand on the discord says "even weirder - just the presence of an unused generic causes this issue" with this [playground link](https://www.typescriptlang.org/play?#code/CYUwxgNghgTiAEAzArgOzAFwJYHtXwiwGcMBVIkGIgClQC55VkBbAI0oBp4McBrEevBIwsqAObwAPvDShEokMACUDAAowczYiAA8Ab3gAoeDIpUGe5FmAMmbToyjMQDYaLEBfANoBdLsfgABygxEAAVPgFXDBFxKRlUOQVgQw8APgBuQ0NQSFgEFHRsPHhrAEZqAA9o2IlpWRB5VEUVIRj3eIamxSzc6DgkNExcfGsAJh0wtKqajvrExuTWtzj5pObgLOyoIgBPdEGikYJiDABBZAwAC3JKGiV4PQCIEAxGEEqMVRDwyMEVuoJdaKeAAXiBiw2WxMYDwJFMdwsXisNjatR8FkczlmcQ8HjBjw8WRMwBwjwCMLhb2QZiIABlTgSoAB3KBYN6EEi3KjUMoABgFXGan2+oQi-FQSmJJgRVAZJAAdDS7grEDgYABRKBgK7UajKmAPUFpckymUGojIsxK6w+AkGinwDxS7JmgD0buFXx+4oEBPK+tp8owCuCYr+UvgHvgzPVvCIjq9ot+Ev9wDGgbuwdDPojGSjbvgrA0EoC+OZVywL3gtA+3vDEqUqUMQA)

and user T6 on discord commented "The reason it's giving that error is CFA
Because the type of nextPageToken depends on usersList, which depends on nextPageToken
Though interestingly it shouldn't really depend on nextPageToken."

### 🙂 Expected behavior

No error.

**Comment by tjjfvi** [comment]
[Minimal repro](https://www.typescriptlang.org/play?#code/FAEwpgxgNghgTmABAMwK4DsIBcCWB7dFPPACgA8AuRAZyzh3QHNEAfRDcZBsEASioAKcPAFsc1MAB5a9JgD5gwGNQCemRCV6IAvHMQBvYIkRQwWRACN4VGQ2ZsOYLuh4726TtxABuI4hB4Bn7GEAS0ljAAXm4wAO4wOObIxCRWcLy+xsZpblaRfgC+iLEAFjimGmm8wAXAQA)

**Comment by MartinJohns** [comment]
Duplicate of #43047.
